### ISSUE TITLE
Differing exception behavior with Task<TResponse> Send and  Task Send

### ISSUE THREAD

**Issue by DavidHall** [jbogard/MediatR#233]
The two send methods in `Mediator.cs` have different behavior when encountering the same exception condition.

The use case is doing the following sort of send:

    mediator.Send(command);

Possibly calling mediator like that is just very bad practice?

The two way method, with the signature:

    public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, CancellationToken cancellationToken = default(CancellationToken))

Throws an exception when it attempts to resolve an unregistered handler.

The one way method, with the signature:

    public Task Send(IRequest request, CancellationToken cancellationToken = default(CancellationToken))

Does not throw - it instead returns a faulted task.

The following tests exercise the differing behavior:

    [Fact]
    public async Task Should_throwasync_for_async_send()
    {
        await Should.ThrowAsync<InvalidOperationException>(async () => await _mediator.Send(new AsyncPing()));
    }

    [Fact]
    public void Should_throw_for_async_send()
    {
        Should.Throw<InvalidOperationException>(_mediator.Send(new AsyncPing()));
    }

    [Fact]
    public async Task Should_throwasync_for_async_void_send()
    {
        await Should.ThrowAsync<InvalidOperationException>(async () => await _mediator.Send(new AsyncVoidPing()));
    }

    [Fact]
    public void Should_throw_for_async_void_send()
    {
        Should.Throw<InvalidOperationException>(_mediator.Send(new AsyncVoidPing()));
    }

I can submit a pull request with these tests.

I was also going to submit a pull request with a "fix" for this, that would look something like the following:

    public Task Send(IRequest request, CancellationToken cancellationToken = default(CancellationToken))
    {
        if (request == null)
        {
            throw new ArgumentNullException(nameof(request));
        }

        var requestType = request.GetType();

        var handler = _voidRequestHandlers.GetOrAdd(requestType,
            t => (RequestHandlerWrapper) Activator.CreateInstance(typeof(RequestHandlerWrapperImpl<>).MakeGenericType(requestType)));

        var task = handler.Handle(request, cancellationToken, _singleInstanceFactory, _multiInstanceFactory);

        if (task.IsFaulted)
        {                
            if (task.Exception != null)
            {
                throw task.Exception;
            }
            throw new InvalidOperationException($"Could not send request of type {typeof(IRequest)}.");           
        }

        return task;
    }

But when I came to authoring tests, I noticed that there are existing `Should.ThrowAsync` tests, so this is at least partially behaving as expected.

**Comment by jbogard** [comment]
So are you not awaiting the task? The expectation is something is awaiting, somewhere.

**Comment by DavidHall** [comment]
No, most of our MediatR calls are made from a synchronous context (webforms codebehind) so with Mediatr 3 we were mostly using the synchronous methods.

I had thought that the `RequestHandler` helper base class was intended for synchronous requests, but perhaps I didn't understand the documentation?

We can change our code to allow async await (by using `RegisterAsyncTask`) but it would be really nice if it was possible to either use MediatR synchronously, or to see the same behavior from the two way and the one way `Send` methods. (it was a nasty surprise when some calls to Mediatr started failing silently, in what looked like an identical use case to calls that threw exceptions).

**Comment by jbogard** [comment]
Yes, MediatR 3.x is async-only in the `IMediator` interface. Once MediatR went async internally, you can really only do an async pipeline (similar to ASP.NET Core, its pipeline is 100% async).

I could have made all the method names "async" I guess? But I would not expose any sync methods so it seemed a bit pointless.

**Comment by DavidHall** [comment]
No the naming is just fine, that really would have been pointless!

Perhaps the documentation needs to be updated or clarified? In the current docs there is the line:

> Or if the request is completely synchronous, inherit from the base RequestHandler class:

That I took to mean that MediatR could be used synchronously.

And a loud call out about the "silent" behavior for one way sends that throw when not awaited, would perhaps save some people heartache. I only caught this when one of my team changed a method to start using Mediatr one way (we usually use two way) and I happened to be using that exact page for something unrelated.

Any way - thanks a lot for the answers, and for the tools. AutoMapper and MediatR have been part of my core toolkit for years now.
